### ISSUE TITLE
bug in printing eObjects

### ISSUE THREAD

**Issue by thibautjombart** [Hackout2/repijson#13]
``` r
> library(repijson)
^[example(toyll)epijson)

toyll> ## Not run: 
toyll> ##D data(toyll)
toyll> ##D toyll
toyll> ##D 
toyll> ##D ind.fields <- c(names(toyll)[1:5], "hospital", "fever", "sleepy")
toyll> ##D 
toyll> ##D x <- as.ejObject(toyll,
toyll> ##D                 recordAttributes=ind.fields,
toyll> ##D                 eventDefinitions=list(
toyll> ##D                 define_ejEvent(name="admission", date="date.of.admission"),
toyll> ##D                 define_ejEvent(name="discharge", date="date.of.discharge"),
toyll> ##D                 define_ejEvent(name="contact1", date="contact1.date", attributes="contact1.id"),
toyll> ##D                 define_ejEvent(name="contact2", date="contact2.date", attributes="contact2.id"),
toyll> ##D                 define_ejEvent(name="contact3", date="contact3.date", attributes="contact3.id")
toyll> ##D                 ))
toyll> ##D x
toyll> ##D 
toyll> ## End(Not run)
toyll> 
toyll> 
toyll> 
> ?toyll
>  data(toyll)
>      toyll
  id    name        dob gender date.of.onset date.of.admission
1  A     Tom 1984-03-14   male    2014-12-28              <NA>
2  B    Andy 1985-11-13   male    2014-12-29        2014-01-05
3 3D Elliiie 1987-06-16 female    2015-01-03        2015-01-12
4 4d     Ana 1987-06-16 female    2015-01-08              <NA>
5  5    Tibo 1982-02-04   male    2015-01-04        2015-01-14
  date.of.discharge    hospital fever sleepy contact1.id contact1.date
1              <NA>        <NA>   yes     no           B    2014-12-26
2              <NA>     St Mary   yes    yes           A    2014-12-26
3        2015-01-17 Whittington    no    yes           5    2014-12-28
4              <NA>        <NA>    no     no        <NA>          <NA>
5        2015-01-17 Whittington   yes    yes          3D    2014-12-28
  contact2.id contact2.date contact3.id contact3.date
1          3D    2014-12-25           B    2014-01-08
2          3D    2014-12-26        <NA>          <NA>
3           5    2014-01-14        <NA>          <NA>
4        <NA>          <NA>        <NA>          <NA>
5           A    2014-12-25        <NA>          <NA>
>      
>      ind.fields <- c(names(toyll)[1:5], "hospital", "fever", "sleepy")
>      
>      x <- as.ejObject(toyll,
+                      recordAttributes=ind.fields,
+                      eventDefinitions=list(
+                      define_ejEvent(name="admission", date="date.of.admission"),
+                      define_ejEvent(name="discharge", date="date.of.discharge"),
+                      define_ejEvent(name="contact1", date="contact1.date", attributes="contact1.id"),
+                      define_ejEvent(name="contact2", date="contact2.date", attributes="contact2.id"),
+                      define_ejEvent(name="contact3", date="contact3.date", attributes="contact3.id")
+                      ))
>      x
Error in attribute$name : $ operator is invalid for atomic vectors
> 
```

**Comment by thibautjombart** [comment]
@twomagpi we're going to try and wrap up examples this pm with @EllieSherrardSmith ; is this issue straightforward to fix? I'm happy to give it a shot if you want - lemme know

**Comment by thibautjombart** [comment]
I've tracked it down to asList_attribute, whose argument was 'NA' and created an error. The function now returns `list(name=NA, type=NA)` when given a NA. @twomagpi is it correct?

**Comment by thibautjombart** [comment]
I now get the following:

``` r
> library(repijson)

Attaching package: ‘repijson’

The following object is masked _by_ ‘.GlobalEnv’:

    toyll

> example(toyll)

toyll> data(toyll)

toyll> toyll
  id    name        dob gender date.of.onset date.of.admission
1  A     Tom 1984-03-14   male    2014-12-28              <NA>
2  B    Andy 1985-11-13   male    2014-12-29        2014-01-05
3 3D Elliiie 1981-04-16 female    2015-01-03        2015-01-12
4 4d     Ana 1977-06-16 female    2015-01-08              <NA>
5  5    Tibo 1982-01-12   male    2015-01-04        2015-01-14
  date.of.discharge    hospital fever sleepy contact1.id contact1.date
1              <NA>        <NA>   yes     no           B    2014-12-26
2              <NA>     St Mary   yes    yes           A    2014-12-26
3        2015-01-17 Whittington    no    yes           5    2014-12-28
4              <NA>        <NA>    no     no        <NA>          <NA>
5        2015-01-17 Whittington   yes    yes          3D    2014-12-28
  contact2.id contact2.date contact3.id contact3.date
1          3D    2014-12-25           B    2014-01-08
2          3D    2014-12-26        <NA>          <NA>
3           5    2014-01-14        <NA>          <NA>
4        <NA>          <NA>        <NA>          <NA>
5           A    2014-12-25        <NA>          <NA>

toyll> ind.fields <- c(names(toyll)[1:5], "hospital", "fever", "sleepy")

toyll> x <- as.ejObject(toyll,
toyll+                 recordAttributes=ind.fields,
toyll+                 eventDefinitions=list(
toyll+                 define_ejEvent(name="admission", date="date.of.admission"),
toyll+                 define_ejEvent(name="discharge", date="date.of.discharge"),
toyll+                 define_ejEvent(name="contact1", date="contact1.date", attributes="contact1.id"),
toyll+                 define_ejEvent(name="contact2", date="contact2.date", attributes="contact2.id"),
toyll+                 define_ejEvent(name="contact3", date="contact3.date", attributes="contact3.id")
toyll+                 ))

toyll> x
{
    "metadata": [

    ],
    "records": [
        {
            "id": 1,
            "attributes": [
                {
                    "name": "id",
                    "type": "character"
                },
                {
                    "name": "name",
                    "type": "character"
                },
                {
                    "name": "dob",
                    "type": "character"
                },
                {
                    "name": "gender",
                    "type": "character"
                },
                {
                    "name": "date.of.onset",
                    "type": "character"
                },
                {
                    "name": "hospital",
                    "type": "character"
                },
                {
                    "name": "fever",
                    "type": "character"
                },
                {
                    "name": "sleepy",
                    "type": "character"
                }
            ],
            "events": [
                {
                    "attributes": [
                        {
                            "name": null,
                            "type": null
                        }
                    ]
                },
                {
                    "attributes": [
                        {
                            "name": null,
                            "type": null
                        }
                    ]
                },
                {
                    "date": "2014-12-26T00:00:00Z",
                    "attributes": [
                        {
                            "name": "contact1.id",
                            "type": "character"
                        }
                    ]
                },
                {
                    "date": "2014-12-25T00:00:00Z",
                    "attributes": [
                        {
                            "name": "contact2.id",
                            "type": "character"
                        }
                    ]
                },
                {
                    "date": "2014-01-08T00:00:00Z",
                    "attributes": [
                        {
                            "name": "contact3.id",
                            "type": "character"
                        }
                    ]
                }
            ]
        },
        {
            "id": 2,
            "attributes": [
                {
                    "name": "id",
                    "type": "character"
                },
                {
                    "name": "name",
                    "type": "character"
                },
                {
                    "name": "dob",
                    "type": "character"
                },
                {
                    "name": "gender",
                    "type": "character"
                },
                {
                    "name": "date.of.onset",
                    "type": "character"
                },
                {
                    "name": "hospital",
                    "type": "character"
                },
                {
                    "name": "fever",
                    "type": "character"
                },
                {
                    "name": "sleepy",
                    "type": "character"
                }
            ],
            "events": [
                {
                    "date": "2014-01-05T00:00:00Z",
                    "attributes": [
                        {
                            "name": null,
                            "type": null
                        }
                    ]
                },
                {
                    "attributes": [
                        {
                            "name": null,
                            "type": null
                        }
                    ]
                },
                {
                    "date": "2014-12-26T00:00:00Z",
                    "attributes": [
                        {
                            "name": "contact1.id",
                            "type": "character"
                        }
                    ]
                },
                {
                    "date": "2014-12-26T00:00:00Z",
                    "attributes": [
                        {
                            "name": "contact2.id",
                            "type": "character"
                        }
                    ]
                },
                {
                    "attributes": [
                        {
                            "name": "contact3.id",
                            "type": "character"
                        }
                    ]
                }
            ]
        },
        {
            "id": 3,
            "attributes": [
                {
                    "name": "id",
                    "type": "character"
                },
                {
                    "name": "name",
                    "type": "character"
                },
                {
                    "name": "dob",
                    "type": "character"
                },
                {
                    "name": "gender",
                    "type": "character"
                },
                {
                    "name": "date.of.onset",
                    "type": "character"
                },
                {
                    "name": "hospital",
                    "type": "character"
                },
                {
                    "name": "fever",
                    "type": "character"
                },
                {
                    "name": "sleepy",
                    "type": "character"
                }
            ],
            "events": [
                {
                    "date": "2015-01-12T00:00:00Z",
                    "attributes": [
                        {
                            "name": null,
                            "type": null
                        }
                    ]
                },
                {
                    "date": "2015-01-17T00:00:00Z",
                    "attributes": [
                        {
                            "name": null,
                            "type": null
                        }
                    ]
                },
                {
                    "date": "2014-12-28T00:00:00Z",
                    "attributes": [
                        {
                            "name": "contact1.id",
                            "type": "character"
                        }
                    ]
                },
                {
                    "date": "2014-01-14T00:00:00Z",
                    "attributes": [
                        {
                            "name": "contact2.id",
                            "type": "character"
                        }
                    ]
                },
                {
                    "attributes": [
                        {
                            "name": "contact3.id",
                            "type": "character"
                        }
                    ]
                }
            ]
        },
        {
            "id": 4,
            "attributes": [
                {
                    "name": "id",
                    "type": "character"
                },
                {
                    "name": "name",
                    "type": "character"
                },
                {
                    "name": "dob",
                    "type": "character"
                },
                {
                    "name": "gender",
                    "type": "character"
                },
                {
                    "name": "date.of.onset",
                    "type": "character"
                },
                {
                    "name": "hospital",
                    "type": "character"
                },
                {
                    "name": "fever",
                    "type": "character"
                },
                {
                    "name": "sleepy",
                    "type": "character"
                }
            ],
            "events": [
                {
                    "attributes": [
                        {
                            "name": null,
                            "type": null
                        }
                    ]
                },
                {
                    "attributes": [
                        {
                            "name": null,
                            "type": null
                        }
                    ]
                },
                {
                    "attributes": [
                        {
                            "name": "contact1.id",
                            "type": "character"
                        }
                    ]
                },
                {
                    "attributes": [
                        {
                            "name": "contact2.id",
                            "type": "character"
                        }
                    ]
                },
                {
                    "attributes": [
                        {
                            "name": "contact3.id",
                            "type": "character"
                        }
                    ]
                }
            ]
        },
        {
            "id": 5,
            "attributes": [
                {
                    "name": "id",
                    "type": "character"
                },
                {
                    "name": "name",
                    "type": "character"
                },
                {
                    "name": "dob",
                    "type": "character"
                },
                {
                    "name": "gender",
                    "type": "character"
                },
                {
                    "name": "date.of.onset",
                    "type": "character"
                },
                {
                    "name": "hospital",
                    "type": "character"
                },
                {
                    "name": "fever",
                    "type": "character"
                },
                {
                    "name": "sleepy",
                    "type": "character"
                }
            ],
            "events": [
                {
                    "date": "2015-01-14T00:00:00Z",
                    "attributes": [
                        {
                            "name": null,
                            "type": null
                        }
                    ]
                },
                {
                    "date": "2015-01-17T00:00:00Z",
                    "attributes": [
                        {
                            "name": null,
                            "type": null
                        }
                    ]
                },
                {
                    "date": "2014-12-28T00:00:00Z",
                    "attributes": [
                        {
                            "name": "contact1.id",
                            "type": "character"
                        }
                    ]
                },
                {
                    "date": "2014-12-25T00:00:00Z",
                    "attributes": [
                        {
                            "name": "contact2.id",
                            "type": "character"
                        }
                    ]
                },
                {
                    "attributes": [
                        {
                            "name": "contact3.id",
                            "type": "character"
                        }
                    ]
                }
            ]
        }
    ]
}

> 
```

**Comment by thibautjombart** [comment]
It looks like the data conversion is going wrong. Most values have been lost. Not sure at which stage this happened.

**Comment by twomagpi** [comment]
At root caused by incorrect type detection in the dataframeToAttributes function utils.R and fixed in the commits leading up to  fbbf810af047d543d71f9581029988ba70518506 
This type of error is now easier to detect by the type checking code in the attribute constructor. (this commit series also fixed something similar in the obkdata conversion routine)

**Comment by thibautjombart** [comment]
Cool!
